packages/store-sync/src/query-cache/queryRECS.ts

```diff
@@ -0,0 +1,118 @@
+  table: T;
+};
+
+type NotQueryFragment<T extends Table> = {
+  type: QueryFragmentType.Not;
+  table: T;
+};
+
+type HasValueQueryFragment<T extends Table> = {
+  type: QueryFragmentType.HasValue;
+  table: T;
+  value: SchemaToPrimitives<T["valueSchema"]>;
+};
+
+type NotValueQueryFragment<T extends Table> = {
+  type: QueryFragmentType.NotValue;
+  table: T;
+  value: SchemaToPrimitives<T["valueSchema"]>;
+};
+
+export function Has<T extends Table>(table: T): HasQueryFragment<T> {
+  return { type: QueryFragmentType.Has, table };
+}
+
+export function Not<T extends Table>(table: T): NotQueryFragment<T> {
+  return { type: QueryFragmentType.Not, table };
+}
+
+export function HasValue<T extends Table>(
+  table: T,
+  value: SchemaToPrimitives<T["valueSchema"]>,
+): HasValueQueryFragment<T> {
+  return { type: QueryFragmentType.HasValue, table, value };
+}
+
+export function NotValue<T extends Table>(
+  table: T,
+  value: SchemaToPrimitives<T["valueSchema"]>,
+): NotValueQueryFragment<T> {
+  return { type: QueryFragmentType.NotValue, table, value };
+}
+
+type QueryFragment<T extends Table> =
+  | HasQueryFragment<T>
+  | NotQueryFragment<T>
+  | HasValueQueryFragment<T>
+  | NotValueQueryFragment<T>;
+
+function fragmentToTableSubject(fragment: QueryFragment<Table>): TableSubject {
+  return {
+    tableId: fragment.table.tableId,
+    subject: Object.keys(fragment.table.keySchema),
+  };
+}
+
+function fragmentToQueryConditions(fragment: QueryFragment<Table>): QueryCondition[] {
+  return Object.entries((fragment as HasValueQueryFragment<Table> | NotValueQueryFragment<Table>).value).map(
+    ([field, right]) => {
+      if (fragment.type === QueryFragmentType.HasValue) {
+        return {
+          left: { tableId: fragment.table.tableId, field },
+          op: "=",
+          right,
+        };
+      } else {
+        return {
+          left: { tableId: fragment.table.tableId, field },
+          op: "!=",
+          right,
+        };
+      }
+    },
+  ) as QueryCondition[];
+}
+
+export async function runQuery<config extends StoreConfig, extraTables extends Tables | undefined = undefined>(
+  store: ZustandStore<AllTables<config, extraTables>>,
+  fragments: QueryFragment<Table>[],
+): Promise<string[]> {
+  const from = fragments
+    .filter(
+      (fragment) =>
+        fragment.type === QueryFragmentType.Has ||
+        fragment.type === QueryFragmentType.HasValue ||
+        fragment.type === QueryFragmentType.NotValue,
+    )
+    .map(fragmentToTableSubject);
+
+  const except = fragments.filter((fragment) => fragment.type === QueryFragmentType.Not).map(fragmentToTableSubject);
+
+  const where = fragments
+    .filter((fragment) => fragment.type === QueryFragmentType.HasValue || fragment.type === QueryFragmentType.NotValue)
+    .map(fragmentToQueryConditions)
+    .flat();
+
+  const result = await query(store, {
+    from,
+    except,
+    where,
+  });
+
+  const entities = result.map((record) => record.join(":"));
```

Comment:
could use `encodeEntity`

but we might consider not using any sort of encoding and how we can just better support composite keys everywhere

Reply:
`encodeEntity` is possible here but a little awkward because the records don't come with their `table` object

How would you support composite keys in RECS without breaking the API?

Reply:
done